src/client/ds3/bucket.c

```diff
@@ -0,0 +1,419 @@
+		dfs_release(dir_obj);
+err_ds3b:
+	ds3_bucket_close(ds3b, ev);
+	return -rc;
+}
+
+int
+ds3_bucket_open(const char *name, ds3_bucket_t **ds3b, ds3_t *ds3, daos_event_t *ev)
+{
+	int           rc;
+	ds3_bucket_t *ds3b_tmp;
+
+	if (ds3 == NULL || name == NULL || ds3b == NULL)
+		return -EINVAL;
+
+	// Prevent attempting to open metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot open metadata bucket");
+		return -ENOENT;
+	}
+
+	D_ALLOC_PTR(ds3b_tmp);
+	if (ds3b_tmp == NULL)
+		return -ENOMEM;
+
+	rc = dfs_connect(ds3->pool, NULL, name, O_RDWR, NULL, &ds3b_tmp->dfs);
+	if (rc != 0) {
+		goto err_ds3b;
+	}
+
+	*ds3b = ds3b_tmp;
+	return 0;
+
+err_ds3b:
+	D_FREE(ds3b_tmp);
+	return -rc;
+}
+
+int
+ds3_bucket_close(ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	int rc = 0;
+
+	rc = dfs_disconnect(ds3b->dfs);
+	D_FREE(ds3b);
+	return -rc;
+}
+
+int
+ds3_bucket_get_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	int               rc       = 0;
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void *const       values[] = {info->encoded};
+	size_t            sizes[]  = {info->encoded_length};
+	daos_handle_t     coh;
+
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	rc = dfs_cont_get(ds3b->dfs, &coh);
+	if (rc != 0) {
+		return -rc;
+	}
+
+	rc = daos_cont_get_attr(coh, 1, names, values, sizes, ev);
+	rc = daos_der2errno(rc);
+	dfs_cont_put(ds3b->dfs, coh);
+	return -rc;
+}
+
+int
+ds3_bucket_set_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	int               rc       = 0;
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void const *const values[] = {info->encoded};
+	size_t const      sizes[]  = {info->encoded_length};
+	daos_handle_t     coh;
+
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	rc = dfs_cont_get(ds3b->dfs, &coh);
+	if (rc != 0) {
+		return -rc;
+	}
+
+	rc = daos_cont_set_attr(coh, 1, names, values, sizes, ev);
+	rc = daos_der2errno(rc);
+	dfs_cont_put(ds3b->dfs, coh);
+	return -rc;
+}
+
+int
+ds3_bucket_list_obj(uint32_t *nobj, struct ds3_object_info *objs, uint32_t *ncp,
+		    struct ds3_common_prefix_info *cps, const char *prefix, const char *delim,
+		    char *marker, bool list_versions, bool *is_truncated, ds3_bucket_t *ds3b)
+{
+	int            rc = 0;
+	char          *file_start;
+	const char    *path        = "";
+	const char    *prefix_rest = prefix;
+	dfs_obj_t     *dir_obj;
+	char          *lookup_path;
+	struct dirent *dirents;
+	daos_anchor_t  anchor;
+	uint32_t       cpi  = 0;
+	uint32_t       obji = 0;
+	uint32_t       i;
+	const char    *name;
+	dfs_obj_t     *entry_obj;
+	mode_t         mode;
+	char          *cpp;
+
+	if (ds3b == NULL || nobj == NULL)
+		return -EINVAL;
+
+	// End
+	if (*nobj == 0)
+		return 0;
+
+	// TODO: support the case when delim is not /
+	if (strcmp(delim, "/") != 0) {
+		return -EINVAL;
+	}
+
+	file_start = strrchr(prefix, delim[0]);
+	if (file_start != NULL) {
+		*file_start = '\0';
```

Comment:
This is a pointer into a const string, correct? Might be better to make a copy rather than modify the original

Reply:
Fixed, see above.